<h3>
Universidade Federal da Paraíba <br>
Centro de Informática <br>
Bacharelado em Ciência de Dados e Inteligência Artificial <br>
Disciplina: Introdução à Inteligência Artificial <br>
Docente: Andrei Formiga <br>
<h3>
<h4>
Discentes:<br> 
Maxwel Barbosa<br>
Vinícius Freitas<br>
Vinícius Vieri <br>
<h4>



<center><h3>Atividade 6 - Algoritmos de busca sem informação<h3></center>

Com o espaço de estados das cidades da Romênia, escolha duas cidades para buscar uma rota usando a implementação de busca em largura que foi vista nos vídeos. Mostre o passo a passo da busca

In [1]:
from busca_bfs import *
from random import randint, seed
seed(5)

In [2]:
#escolhendo a origem e o destino da rota
i = randint(0, len(estados_romenia)-1)
j = randint(0, len(estados_romenia)-1)

print(i, j)

19 8


In [3]:
origem = estados_romenia[i]['estado']
origem

'Neamt'

In [4]:
destino = estados_romenia[j]['estado']
destino

'Craiova'

- No caso do nosso problema a busca começa na cidade Neamt e precisa chegar até a cidade de Craiova.
- Iremos utilizar o método de busca em largura (Breadth First Search) como método para percorrer a árvore.
- Iremos executar os passos de busca (criação e análise dos nós da árvore de busca) até chegar ao objetivo desejado.

In [5]:
#uma vez definidos o ponto de partida e destino da rota, inicia-se a busca
no_neamt = No('Neamt', 0, None, None)

prob_neamt_craiova = Problema(estados_romenia, no_neamt, lambda no: no.estado == destino)

In [6]:
bfs_neamt_craiova = BuscaLargura(prob_neamt_craiova)
bfs_neamt_craiova.fronteira

[(Neamt, 0)]

Inicia-se a busca a partir de Neamt <br>
Um passo-a-passo do algoritmo é exibido abaixo até chegar no objetivo e destino final, Craiova

In [7]:
bfs_neamt_craiova.passo_busca()
bfs_neamt_craiova.fronteira

[(Iasi, 87)]

In [8]:
bfs_neamt_craiova.passo_busca()
bfs_neamt_craiova.fronteira

[(Vaslui, 179)]

In [9]:
bfs_neamt_craiova.passo_busca()
bfs_neamt_craiova.fronteira

[(Urziceni, 321)]

In [10]:
situacoes_finais = [BUSCA_SUCESSO, BUSCA_FALHOU]

while bfs_neamt_craiova.situacao not in situacoes_finais:
    bfs_neamt_craiova.passo_busca()
    print(bfs_neamt_craiova.fronteira)

[(Bucharest, 406), (Hirsova, 419)]
[(Hirsova, 419), (Fagaras, 617), (Pitesti, 507), (Giurgiu, 496)]
[(Fagaras, 617), (Pitesti, 507), (Giurgiu, 496), (Eforie, 505)]
[(Pitesti, 507), (Giurgiu, 496), (Eforie, 505), (Sibiu, 716)]
[(Giurgiu, 496), (Eforie, 505), (Sibiu, 716), (Rimnicu Vilcea, 604), (Craiova, 645)]
[(Eforie, 505), (Sibiu, 716), (Rimnicu Vilcea, 604), (Craiova, 645)]
[(Sibiu, 716), (Rimnicu Vilcea, 604), (Craiova, 645)]
[(Rimnicu Vilcea, 604), (Craiova, 645), (Arad, 856), (Oradea, 867)]
[(Craiova, 645), (Arad, 856), (Oradea, 867)]
[(Arad, 856), (Oradea, 867)]


In [11]:
bfs_neamt_craiova.passo_busca()
bfs_neamt_craiova.fronteira

Busca chegou ao objetivo com sucesso


[(Arad, 856), (Oradea, 867)]

In [12]:
bfs_neamt_craiova.solucao

[(Neamt, 0),
 (Iasi, 87),
 (Vaslui, 179),
 (Urziceni, 321),
 (Bucharest, 406),
 (Pitesti, 507),
 (Craiova, 645)]

A solução encontrada pelo algoritmo de busca em largura obteve sucesso com o valor da solução 645.

In [13]:
custo_total = bfs_neamt_craiova.solucao[-1].custo
custo_total

645